src/arch/riscv/rvdbg013.c

```diff
@@ -0,0 +1,1772 @@
+{
+	RVDBGv013_DMI_t *dmi = t->priv;
+	int res;
+	/* Try Hartreset first*/
+	res = rvdbg_dmi_write(dmi, DMI_REG_DMCONTROL, DMCONTROL_DMACTIVE | DMCONTROL_HARTRESET);
+	if (res)
+		DEBUG_WARN("Reset write HARTRESET failed\n");
+	uint32_t dmcontrol = 0;
+	res = rvdbg_dmi_read(dmi, DMI_REG_DMCONTROL, &dmcontrol);
+	if (res)
+		DEBUG_WARN("Reset read dmcontrol failed\n");
+	if (!(dmcontrol & DMCONTROL_HARTRESET)) {
+		DEBUG_WARN("Optional HARTRESET not implemented, using NDMRESET\n");
+		res = rvdbg_dmi_write(dmi, DMI_REG_DMCONTROL, DMCONTROL_DMACTIVE | DMCONTROL_NDMRESET);
+		if (res)
+			DEBUG_WARN("Reset write NDMRESET failed\n");
+	}
+	res = rvdbg_dmi_write(dmi, DMI_REG_DMCONTROL, DMCONTROL_DMACTIVE);
+	if (res)
+		DEBUG_WARN("Reset release RESET failed\n");
+}
+
+static void rvdbg_halt_resume(target *t, bool step)
+{
+	RVDBGv013_DMI_t *dmi = t->priv;
+	int res;
+	/* Handle single step in DCSR*/
+	uint32_t dcsr;
+	res = rvdbg_read_single_reg(dmi, HART_REG_CSR_DCSR, &dcsr, AUTOEXEC_STATE_NONE);
+	if (res) {
+		DEBUG_WARN("Halt_resume read DCSR failed\n");
+	} else {
+		DEBUG_TARGET("DCSR start 0x%08" PRIx32 "\n", dcsr);
+		if (step)
+			dcsr |= CSR_DCSR_STEP;
+		else
+			dcsr &= ~CSR_DCSR_STEP;
+		res = rvdbg_write_single_reg(dmi, HART_REG_CSR_DCSR, dcsr, AUTOEXEC_STATE_NONE);
+		if (res)
+			DEBUG_WARN("Write DCSR failed\n");
+	}
+	uint32_t dmstatus;
+	/* As DMCONTROL_HALTREQ is not set, this clears halt request*/
+	if (rvdbg_dmi_write(dmi, DMI_REG_DMCONTROL, DMCONTROL_DMACTIVE| DMCONTROL_RESUMEREQ) < 0) {
+		DEBUG_WARN("Can not write resumereq\n");
+		dmi->error = true;
+	}
+	dmstatus = 0;
+	platform_timeout timeout;
+	platform_timeout_set(&timeout, 1050); /* Hart should resume in less than 1 sec*/
+	while (!DMSTATUS_GET_ALLRESEUMSET(dmstatus)) {
+		if (rvdbg_dmi_read(dmi, DMI_REG_DMSTATUS, &dmstatus) < 0) {
+			DEBUG_WARN("Can not read dmstatua\n");
+			dmi->error = true;
+		}
+		if (platform_timeout_is_expired(&timeout)) {
+			DEBUG_WARN("Timeout waiting for resume, dmstatus 0x%08" PRIx32 "\n", dmstatus);
+			dmi->error = true;
+			return;
+		}
+	}
+	if (rvdbg_dmi_write(dmi, DMI_REG_DMCONTROL, DMCONTROL_DMACTIVE) < 0) {
+		DEBUG_WARN("Can not write resumereq\n");
+		dmi->error = true;
+	}
+}
+
+static enum target_halt_reason rvdbg_halt_poll(target *t, target_addr *watch)
+{
+	(void)watch;
+ 	RVDBGv013_DMI_t *dmi = t->priv;
+	int res;
+	uint32_t dmstatus;
+	res = rvdbg_dmi_read(dmi, DMI_REG_DMSTATUS, &dmstatus);
+	if (res)
+		DEBUG_WARN("Poll write dmcontrol failed\n");
+	if (!DMSTATUS_GET_ALLHALTED(dmstatus))
+		return TARGET_HALT_RUNNING;
+	/* dcsr may not be readable when running*/
+	uint32_t dcsr;
+	res = rvdbg_read_single_reg(dmi, HART_REG_CSR_DCSR, &dcsr, AUTOEXEC_STATE_NONE);
+	uint8_t cause = (dcsr >> 6) & 7;
+	if ((t->t_designer == 0x612) && (cause == 3 ) && (dcsr & CSR_DCSR_STEP)) {
+		/* FIXME: ESP32-C3 never reports TARGET_HALT_STEPPING */
+		DEBUG_INFO("Workaround for single stepping ESP32-C3\n");
```

Comment:
is it possible to identify the esp32-c3? maybe implement this explicitly as a quirk, makes it more visible, it is a target specific workaround. Other hw should probably still be handled normally.

Reply:
It is already found:
...
Machine 612, 80000001, 1, 0
Found 8 triggers
***  1      ESP32-C3 Generic RVDBG 0.13